nvm.sh

```diff
@@ -2280,6 +2280,12 @@ nvm_download_artifact() {
     return 4
   )
 
+  # Check if the file is successfully downloaded
+  if [ ! -f "${TARBALL}" ]; then
+      nvm_err 'Download file not found, skip checksum.'
```

Comment:
it'd be nice if this error message explained a bit more about what could cause this - also, that it's skipping the checksum is sort of incidental.

Reply:
There is a solution that we should fix nvm_download() method to show the reason of fail. The nvm_download method is used too much, and I'm concerned that the impact of the change is too large. I found "if nvm_grep '404 Not Found' "${TARBALL}" >/dev/null; then" code in the nvm_download_artifact, so add before it. I will take another pull request to change nvm_download().  Let's compare and see which solution is better.

Reply:
I'm just suggesting changing this one line - why would the downloaded file not be found?

Reply:
May i change this line to "nvm_err 'Binary download failed, skip checksum.'" ?

Reply:
I don’t think it should talk about the checksum at all; that’s something the user doesn’t care about - it’s the bug you’re fixing, but the issue is that there’s no downloaded file. Let’s just tell them that, and ideally, why it’s potentially not there.

Reply:
The reason of the file download failed is that there is no file in the npm warehouse. Do you mean that I should remind users to check whether there is such a file in the npm warehouse? But this doesn't seem very reasonable, it is also possible that the file cannot be downloaded due to network problems.

Reply:
If by "npm warehouse" you mean "nvm cache", then sure, but i mean, it just tried to download it - why isn't it there? Did the download get interrupted, or is the disk full, or something else?

Reply:
"npm warehouse" which is store install file of npm, like "https://registry.npmjs.org/", "http://registry.mirror.cqupt.edu.cn/", "https://registry.npm.taobao.org/" etc. For the speed of download, we always create a mirror warehouse to proxy the official warehouse. When the files in the official warehouse are updated, but the files in the mirror warehouse are not updated, the download will fail. The problem I'm having is that the file named index.tab have been updated, but the version of npm file is not update in npm warehouse.

Reply:
I don't think so, as a user needs to know why the download failed, in order to make the next improvement. I found that the nvm.sh file has included a check for 404, but if a 404 occurs, it does not enter the method. Do we modify this piece of code to execute this function when the download fails?

my test code as following.
```shell
  nvm_err "Downloading ${TARBALL_URL}..."
  nvm_download -L -C - "${PROGRESS_BAR}" "${TARBALL_URL}" -o "${TARBALL}" || (
    command rm -rf "${TARBALL}" "${tmpdir}"
    nvm_err "Binary download from ${TARBALL_URL} failed, trying source."
    return 4
  )

  nvm_err "404 check begin"
  if nvm_grep '404 Not Found' "${TARBALL}" >/dev/null; then
    command rm -rf "${TARBALL}" "${tmpdir}"
    nvm_err "HTTP 404 at URL ${TARBALL_URL}"
	nvm_err "404 checking..."
    return 5
  fi
```
Test result
![image](https://user-images.githubusercontent.com/22415690/232010383-5e1dec75-3844-4135-aa5e-2066a51723b3.png)

Reply:
It certainly shouldn't enter the method with a 404, but i had to add that because sometimes it provides a code of 200 and content that indicates it's a 404 :-)

Reply:
What do you think about checking whether the download is success after returning the result in the "nvm_download " function? In fact, "return 4" in the download function does not work. If it does, the function "nvm_compare_checksum" will not excute if the file cannot be downloaded.

```shell
nvm_err "Downloading ${TARBALL_URL}..."
  nvm_download -L -C - "${PROGRESS_BAR}" "${TARBALL_URL}" -o "${TARBALL}" || (
    command rm -rf "${TARBALL}" "${tmpdir}"
    nvm_err "Binary download from ${TARBALL_URL} failed, trying source."
    return 4---it does not work, function not return.
  )
```

Reply:
oh hm, i wonder why that doesn't work? maybe because it's in the subshell?

Reply:
yes, So I think it should be fine to fix this return.

Reply:
Sounds good to me.